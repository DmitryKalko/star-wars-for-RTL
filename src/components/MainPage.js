import React, { useEffect, useState } from "react";
import axios from "axios";
import './App.css';
import FiguresList from './FiguresList';

// import { Chart as ChartJS, ArcElement, Tooltip, Legend, LineController, LineElement, PointElement, LinearScale, Title,CategoryScale } from 'chart.js';
// import { Doughnut } from 'react-chartjs-2';
// import { Line } from 'react-chartjs-2';


// ChartJS.register(ArcElement, Tooltip, Legend, LineController, LineElement, PointElement, LinearScale, Title, CategoryScale);
// const data = {
//     labels:['Red', 'Blue', 'Yellow', 'Green', 'Purple', 'Orange'],
//     datasets: [{
//         label: '# of Votes',
//         data: [12, 19, 3, 5, 2, 3],
//         backgroundColor: [
//           'rgba(255, 99, 132, 0.2)',
//           'rgba(54, 162, 235, 0.2)',
//           'rgba(255, 206, 86, 0.2)',
//           'rgba(75, 192, 192, 0.2)',
//           'rgba(153, 102, 255, 0.2)',
//           'rgba(255, 159, 64, 0.2)',
//         ],
//         borderColor: [
//           'rgba(255, 99, 132, 1)',
//           'rgba(54, 162, 235, 1)',
//           'rgba(255, 206, 86, 1)',
//           'rgba(75, 192, 192, 1)',
//           'rgba(153, 102, 255, 1)',
//           'rgba(255, 159, 64, 1)',
//         ],
//         borderWidth: 1,
//     }]
// }


const MainPage = () => {

    const baseUrl = 'https://react-test-starwars.vercel.app/api/products';

    const [figuresData, setFiguresData] = useState([]);
    const [currentPage, setCurrentPage] = useState(0);
    const [fetching, setFetching] = useState(true);
    const [totalPages, setTotalPages] = useState(1);

    useEffect(() => {
        if (fetching && currentPage < totalPages) {
            let newCurrentPage = currentPage + 1;
            setCurrentPage(newCurrentPage);
            axios.get(`${baseUrl}?page=${newCurrentPage}`)
                .then((response) => {
                    setFiguresData([...figuresData, ...response.data.data])
                    setTotalPages(response.data.meta.totalPages)
                })
                .finally(() => setFetching(false))
        }
    }, [fetching]);

    useEffect(() => {
        document.addEventListener('scroll', scrollHandler);
        return function () {
            document.removeEventListener('scroll', scrollHandler);
        };
    }, []);

    const scrollHandler = (e) => {
        if (e.target.documentElement.scrollHeight - (e.target.documentElement.scrollTop + window.innerHeight) < 600) {
            setFetching(true)
        };
    }

    return (
        <div className='wrapper'>
            <h1 className="main-title">Star Wars <br /> Figures</h1>
            <p className="subtitle">Find the latest products for the biggest fans <br /> of the iconic saga.</p>
            {/* <div style={{ width: '300px', margin: '0 auto' }}>
                <Doughnut data={data} />
            </div>
            <Line
                datasetIdKey='id'
                data={{
                    labels: ['Jun', 'Jul', 'Aug'],
                    datasets: [
                        {
                            id: 1,
                            label: '',
                            data: [5, 6, 7],
                        },
                        {
                            id: 2,
                            label: '',
                            data: [3, 2, 1],
                        },
                    ],
                }}
            /> */}

            {figuresData && (
                <FiguresList data-testid="FiguresList" figuresData={figuresData} />
            )}
        </div>
    );
};

export default MainPage;

