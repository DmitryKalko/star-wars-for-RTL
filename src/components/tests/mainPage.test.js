import React from 'react';
import axios from 'axios';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom'
import { userEvent } from '@testing-library/user-event';
import MainPage from '../MainPage';

// ОПИСАНИЕ
// ОТЛИЧИЯ RTL ОТ ENZYME
// - React Testing Library работает напрямую с реальными ДОМ узлами, а Enzyme работает с экземплярами отрендоренных компонентов
// - RTL не опирается на классы, айди или теги элементов, а опираентся на их содержание (эмитирует поведение пользователя)
// - снимки можно делать но по сути они не нужны поскольку мы не опираемся на внутреннюю реализацию компонента (т е при render мы увидим тоже самое что и при снапшоте)
// - используются регулярки для написания более гибких тестов (что бы не проверять точное совпадения по тексту)





// чем отличается fireEvent от userEvents?



//jest.mock("axios");  //  замокали либу аксиос

const goods = [
    {
        "id": 1,
        "shortDescription": "Veteran of the Galactic Civil War who fought under the banner of the Rebellion.",
        "image": "/star-wars/img-01.jpg",
        "name": "Cara Dune",
        "price": 1599
    },
    {
        "id": 2,
        "shortDescription": "She's a force-sensitive human who became Jedi after years of living as scavenger.",
        "image": "/star-wars/img-02.jpg",
        "name": "Rey Skywalker",
        "price": 1599
    },
    {
        "id": 3,
        "shortDescription": "Imperial service. Gideon is a human who serves as an officer in the Imperial Security",
        "image": "/star-wars/img-03.jpg",
        "name": "Moff Gideon",
        "price": 1599
    },
];


describe('MainPage', () => {

    it("should have title", () => {
       const { getByText } = render(<MainPage/>);   // getByText и другие, это методы объекта screen
       const titleElement = getByText(/Star Wars/i);
       expect(titleElement).toBeInTheDocument();

    //    render(<MainPage/>);
    //    screen.debug()  // покажет разметку страницы как ее увидит пользователь

    // const { asFragment } = render(<MainPage/>);
    // expect(asFragment(<MainPage/>)).toMatchSnapshot()
    });


    // it("fetch new data", async () => {
    //     axios.get.mockImplementationOnce(() => Promise.resolve({data: {goods}}));
    //     // const { getByTestId } = render(<MainPage/>);
    //     // const items = await getByTestId('FiguresList');
    //     // expect(items).toHaveLength(3)


    //     expect(axios.get).toBeCalledTimes(1);
    //     expect(axios.get).toBeCalledWith("https://react-test-starwars.vercel.app/api/products");
    // })
});