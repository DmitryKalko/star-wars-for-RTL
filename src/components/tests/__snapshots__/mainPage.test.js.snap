// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage should have title 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <h1
      class="main-title"
    >
      Star Wars 
      <br />
       Figures
    </h1>
    <p
      class="subtitle"
    >
      Find the latest products for the biggest fans 
      <br />
       of the iconic saga.
    </p>
    <div
      class="figuresContainer"
    />
  </div>
</DocumentFragment>
`;
